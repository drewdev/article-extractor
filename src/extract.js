export function extract(content) {
  let matchedElements;
  let scrubbedElements;
  let targetedContent;
  let extractedArticle;

  const elementsToSelect = ['article','section'];
  const tagsToScrub = ['style','noscript','script','iframe','footer','gpt-ad'];
  const sortByInnerHTML = (a,b) => (a.innerHTML.length < b.innerHTML.length) ? 1 : -1;

  function scrub(tagName, raw) {
    if (raw.indexOf(`<${tagName}`) > 0) {
      const openingSelector = `<${tagName}`;
      const closingSelector = `</${tagName}>`;
      const startIndex = raw.indexOf(openingSelector);
      const endIndex = raw.indexOf(closingSelector) + closingSelector.length - 1;

      let transformed = raw.substring(0,startIndex) + raw.substring(endIndex + 1, raw.length)
      return scrub(tagName, transformed)
    } else {
      return raw;
    }
  }

  matchedElements = elementsToSelect.reduce((accumulator, tag) => {
    return accumulator.concat(Array.from(content.getElementsByTagName(tag)));
  }, []);


  if (matchedElements.length > 0) {
    scrubbedElements = matchedElements.reduce((elements, element) => {
      elements.push({
        tagName: element.tagName,
        innerHTML: tagsToScrub.reduce((accumulator, tagToScrub) => scrub(tagToScrub, accumulator), element.innerHTML)
      });
      return elements
    }, []);

    targetedContent = scrubbedElements.sort(sortByInnerHTML)[0];

    extractedArticle = `<${targetedContent.tagName}>${targetedContent.innerHTML}</${targetedContent.tagName}>`;

    content.body.innerHTML = extractedArticle;
  }

  return content
}
