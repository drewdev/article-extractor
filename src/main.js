import {extract} from './article-extractor';

extract(document);
