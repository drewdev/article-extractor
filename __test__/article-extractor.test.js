import React from 'react';
import ReactDOM from 'react-dom';
import {extract} from '../src/extract';
import NestedArticleWithMixedScrubbableTagsFixture from './fixtures/NestedArticleWithMixedScrubbableTagsFixture.react'

beforeEach(() => {
  document.body.innerHTML = '';
  let div = document.createElement('div')
  div.setAttribute('id', 'container');
  document.body.appendChild(div);
})

test('extracts nested article with embedded mixed tags to scrub', () => {
  ReactDOM.render(<NestedArticleWithMixedScrubbableTagsFixture />, document.getElementById('container'));
  expect(extract(document).body.innerHTML).toMatchSnapshot();
});
