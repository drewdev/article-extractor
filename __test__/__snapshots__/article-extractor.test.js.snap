// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extracts nested article with embedded mixed tags to scrub 1`] = `"<article><div><header>THIS IS THE HEADER OF THE ARTICLE</header></div><div><section><p>Here is some article text that we want to match</p></section></div><div><div></div></div><div><section><p>Here is some more article text that we want to match</p></section></div><div></div></article>"`;
